### PR TITLE
Add ActiveEntities support

### PR DESCRIPTION
**Pull request by juanshishido**
TODO

* ✅ Add an example script to [/examples](https://github.com/twitterdev/twitter-python-ads-sdk/tree/master/examples) (see [this](https://github.com/twitterdev/twitter-python-ads-sdk/pull/180/files#diff-f1aa2241650a1fee44b3d8854b6bae5e))
* <strike>Rebase after #179 is merged</strike>
    * ✅ Use `API_VERSION`

---

Quick example.

[**EDIT**] See the example file instead.

### CODE REVIEW

File: twitter_ads/active_entities.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+from twitter_ads.enum import TRANSFORM
+from twitter_ads.resource import resource_property, Resource
+
+
+class ActiveEntities(Resource):
+
+    PROPERTIES = {}
+
+    # TODO: use `API_VERSION`
```
Comment (tushdante) [inline twitter_ads/active_entities.py]: I know this is listed as a TODO but could you make the change to `API_VERSION`? We can assume this won't be merged until #179 is

**Comment by tushdante** [comment]
Nevermind, I see that you've already listed all the TODOs :)

**Comment by juanshishido** [comment]
I'm wondering if something like this would be more intuitive and in line with what's already in place with analytics

```
>>> LineItem.active_entities(account, start_time, end_time)
[
  {
    "activity_end_time": "2019-03-05T01:59:18Z",
    "activity_start_time": "2019-03-05T01:59:18Z",
    "entity_id": "eelsr"
  },
  {
    "activity_end_time": "2019-03-05T10:10:50Z",
    "activity_start_time": "2019-03-05T09:07:03Z",
    "entity_id": "eeukx"
  },
  {
    "activity_end_time": "2019-03-05T10:03:32Z",
    "activity_start_time": "2019-03-05T09:40:21Z",
    "entity_id": "eeulb"
  },
  {
    "activity_end_time": "2019-03-05T14:40:59Z",
    "activity_start_time": "2019-03-05T11:37:30Z",
    "entity_id": "eeulr"
  }
]
```

This seems cleaner to me. Thoughts, @twitterdev/ads-api-sdks?

**Comment by jbabichjapan** [comment]
> I'm wondering if something like this would be more intuitive and in line with what's already in place with analytics
> 
> ```
> >>> LineItem.active_entities(account, start_time, end_time)
> [
>   {
>     "activity_end_time": "2019-03-05T01:59:18Z",
>     "activity_start_time": "2019-03-05T01:59:18Z",
>     "entity_id": "eelsr"
>   },
>   {
>     "activity_end_time": "2019-03-05T10:10:50Z",
>     "activity_start_time": "2019-03-05T09:07:03Z",
>     "entity_id": "eeukx"
>   },
>   {
>     "activity_end_time": "2019-03-05T10:03:32Z",
>     "activity_start_time": "2019-03-05T09:40:21Z",
>     "entity_id": "eeulb"
>   },
>   {
>     "activity_end_time": "2019-03-05T14:40:59Z",
>     "activity_start_time": "2019-03-05T11:37:30Z",
>     "entity_id": "eeulr"
>   }
> ]
> ```
> This seems cleaner to me. Thoughts, @twitterdev/ads-api-sdks?

It does look cleaner, but since you're more likely to have a loop that loops through a bunch of different entity types and calls one by one it might be fine to start it as the active_entities.all() one (because you could switch between different entity types without breaking out of a single for loop or chunk of operation), if you really want to see the difference you would have to write both all the way to sample level...also for non-API implementation it's more likely to look like active_entities.all() one, the way we have wrapped analytics calls into the entity objects is one of the most additive things about the SDK.

**Comment by juanshishido** [comment]
Thanks for the feedback, @jbabichjapan.

>The way we have wrapped analytics calls into the entity objects is one of the most additive things about the SDK

I agree and this is why I think we should follow this pattern. By adding an `active_entities` method under the `Analytics` class, we're able to enable users to leverage the workflows they use for analytics. Analytics requests are made by calling an analytics method on a class: `PromotedTweet.all_stats()`, for example. With the direction I'm proposing, the flow would be:

```
>>> active_entities = PromotedTweet.active_entities(...)  # and extract entity IDs and dates
>>> PromotedTweet.all_stats(...)  # passing in IDs and start and end times from the active entities call as well as other request parameters
```

I believe that adding this under the `Analytics` class makes sense in terms of organization. Other classes&mdash;like `Campaign` or `LineItem`&mdash;then don't need to inherit from yet another class.

Some of the reasons I think it doesn't make sense to have an `ActiveEntities` class are:

1. If we're inheriting from `Resource`, then I have to override several of the methods (as the PR currently shows)
    * Active Entities isn't really a resource like campaigns or line items are; there's no index and show, for example
2. Having an `ActiveEntities` class that returns a `Cursor` object could encourage undesirable practices&mdash;i.e., pulling analytics for a single entity ID at a time

Let me put up the PR with the example. Being able to see it could help clarify.

Thanks again.

File: examples/active_entities.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+# Active Entities for Line Items
+active_entities = LineItem.active_entities(account, start_time, end_time)
+
+# Entity IDs to fetch analytics data for
+# Note: analytics endpoints support a
+# maximum of 20 entity IDs per request
+ids = entity_ids(active_entities)
+
+# Date range for analytics request
+start, end = date_range(active_entities, fetch_frequency='DAY')
```
Comment (tushdante) [inline examples/active_entities.py]: nit but does this need to be `fetch_frequency` or just `frequency`?

File: twitter_ads/resource.py [hunk 3]
```diff
@@ -218,13 +218,16 @@ class Analytics(object):
...
+        'FundingInstrument': ENTITY.FUNDING_INSTRUMENT,
         'LineItem': ENTITY.LINE_ITEM,
+        'MediaCreative': ENTITY.MEDIA_CREATIVE,
         'OrganicTweet': ENTITY.ORGANIC_TWEET,
         'PromotedTweet': ENTITY.PROMOTED_TWEET
     }
 
     RESOURCE_SYNC = '/' + API_VERSION + '/stats/accounts/{account_id}'
     RESOURCE_ASYNC = '/' + API_VERSION + '/stats/jobs/accounts/{account_id}'
+    RESOURCE_ACTIVE_ENTITIES = '/5/stats/accounts/{account_id}/active_entities'
```
Comment (tushdante) [inline twitter_ads/resource.py]: change this to `API_VERSION`

File: twitter_ads/resource.py [hunk 4]
```diff
@@ -307,3 +310,21 @@ def async_stats_job_data(klass, account, url, **kwargs):
...
 
         return response.body
+
+    @classmethod
+    def active_entities(klass, account, start_time, end_time, **kwargs):
+        entity_type = klass.__name__
+        if entity_type == 'OrganicTweet':
+            raise ValueError("'OrganicTweet' not support with 'active_entities'")
+
+        # TODO: It's important that the start end end times be expressed in
```
Comment (tushdante) [inline twitter_ads/resource.py]: nit: maybe change this to `NOTE` or something?

File: twitter_ads/utils.py [hunk 5]
```diff
@@ -55,3 +67,26 @@ def size(default_chunk_size, response_time_max, response_time_actual):
     scale = 1 / (response_time_actual / response_time_max)
     size = int(default_chunk_size * scale)
     return min(max(size, 1), default_chunk_size)
+
+
+def entity_ids(data):
```
Comment (tushdante) [inline twitter_ads/utils.py]: is this necessary? rather, does this need to be part of the SDK or just something we expect users to handle?

**Comment by juanshishido** [comment]
Discussed with @tushdante and I have the following TODOs:

1. Revert 94202a7199e13844608776cd9edfe66b94749cde, which is the original implementation where there was an `ActiveEntities` class and a method under the `Account` class. This is being removed in favor of the latest implementation which adds `active_entities` as a method under the `Analytics` class (as shown in the example).
2. Will use `API_VERSION` version for all URLs.
3. Will undo some of the changes in bf7eee55c65335b6aeb6a690b1d207433d4c6e73. Specifically, I will remove the `entity_ids` and `date_ranges` functions. The reasoning&mdash;for the latter specifically&mdash;is that we don't want to couple analytics fetching logic with the active entities method.

**Pull request merged**